### ISSUE TITLE
It is possible to place nodes inside the player even with the setting disabled

### ISSUE THREAD

**Issue by LuciusofLegend** [minetest/minetest#12744]
##### Minetest version

```
Minetest 5.6.0
```

##### OS / Hardware

```
Operating system: Linux (Manjaro KDE 21.3.7)
CPU: Intel i5-10500H
```

##### Summary
I have the 'build inside player' setting disabled, but I can still place nodes inside of myself.

##### Steps to reproduce
Install and run Minetest.
Select Minetest Dev.
Create and enter a new world.
Place any kind of node at eye/legs level, while moving back and forth with the w and s keys.
See how your head goes inside the node sometimes.  This also happens while building straight up, your feet sink into one node.
The nodes can still be interacted with as normal, and once you more out of them, you cannot move back in.
This bug is present in both survival and creative mode.

**Comment by appgurueu** [comment]
Minetest presumably doesn't do a proper box-box collision check but rather just checks whether the player position (feet position) would be in the node, hence the eye position being in the node is not detected.

Similarly, I expect the setting to fail for nodes with unusual (read: not 1³m) collisionboxes.

**Comment by Zughy** [comment]
@appgurueu I've tested with nodes on legs position and I can still reproduce it. I was actually too lazy to come up with a way to test the eye position case

**Comment by Desour** [comment]
Relevant code:
https://github.com/minetest/minetest/blob/643971c948c1aef424f2a55e81414e30958c40ee/src/client/game.cpp#L3527-L3536

* (`walkable` means that objects collide, btw., you can walk on it. And `n` is the (predicted) placed node.)
* It's just a client-side check. It would make much more sense to do this on the server, i.e. in `core.item_place_node`.
* The check for noclip looks like it should check for noclip *and fly*.
* (The 2nd check for `nodedef->get(n).walkable` is always evaluated to `true` if it is evaluated, assuming nodedefs don't spontaneously change.)
* `neighbourpos` is the `pointed_thing.above`. So for whatever reason, this doesn't take `buildable_to` into account.
* (Btw. `getStandingNodePos` and much other code assumes that player (and object) positions are at the ground. Maybe we should make this stuff use the collisionbox at some point.)
* The last two lines in the condition assume the player is 2 nodes high. And they don't take node or player collision boxes into account.

**Comment by appgurueu** [comment]
> It's just a client-side check. It would make much more sense to do this on the server.

[`disable_build_where_they_stand`](https://github.com/appgurueu/disable_build_where_they_stand) implements everything correctly serverside, as a mod. It obviously suffers from poor client prediction.

**Comment by LuciusofLegend** [comment]
I just want to clarify something that may not have been clear before.  Did you notice while recreating that you can stay inside the node?
![Bug](https://user-images.githubusercontent.com/83921153/188874838-19e55844-9bb0-4f32-b402-7f4da14a97f0.png)
![Bug2](https://user-images.githubusercontent.com/83921153/188874852-bd67a5a0-abe5-4633-8e20-3506da310485.png)

Also, I would like to say, I think box-box collision would be the most versatile solution.  This would allow modders to, say, make the player larger or smaller, as mentioned above.  Unfortunately, I don't know enough c++ to write the code myself, but it looks to me like the best method long term.

**Comment by joleman11** [comment]
Instead of only taking players into account, there should be a customizable option per entity type that determines if you are allowed to place nodes where they are standing.

**Comment by appgurueu** [comment]
> Instead of only taking players into account, there should be a customizable option per entity type that determines if you are allowed to place nodes where they are standing.

No, not at all. Currently there is a client-side player setting. There are efforts to make this server-sent to some extent however. There is no point in extending this to entities; if mod authors don't want their entities to place nodes where they are at, they have to implement it themselves - it is not the business of the engine.

**Comment by Desour** [comment]
^ I think the comment was referring to whether one can place nodes inside the object, not whether the object can place nodes in itself.
Such an obj property sounds reasonable, imo.
This would also mean that you then can't place nodes inside other players, btw.. (Idk if I'd like this.)
